### ISSUE TITLE
gazelle turns dep on :dogproto in //path/to/dir into //path/to/dir/dogproto:go_default_library

### ISSUE THREAD

**Issue by jmhodges** [bazelbuild/rules_go#280]
In the directory `path/to/dir` with this target

    go_proto_library(
        name = "dogproto",
        srcs = ["dog.proto"],
    )

and this target depending on it

    go_library(
        name = "go_default_library",
        srcs = ["main.go",]
        deps = [":dogproto"],
    )

running gazelle of the BUILD file containing it will turn the ":dogproto" in the deps into `//path/to/dir/dogproto:go_default_library` which is a target that does not exist.

This is not specific to go_proto_library. It appears to also happen when the dependency is a `go_library` that has sources in a subdirectory. (gazelle errors out in a different way if there are multiple packages in the Go code files that are in the same directory.)

**Comment by jmhodges** [comment]
This is with the rules_go at commit fb1500ed88cc8d49f77b47dd0351275ebd4f4003

**Comment by jmhodges** [comment]
Seems it also has the same bug if the deps is `["//path/to/dir:dogproto"]`.

**Comment by jmhodges** [comment]
I'd suspect gazelle is getting this path from the import paths and not correcting for the rules its already read. This gets hidden in the `go_library` with a subdirectory case because gazelle is creating a BUILD file in that subdirectory that has a matching target (and because gazelle doesn't support more than one Go package in a directory).

While, in that case, I could imagine this being closed as a design choice, it being broken in the `go_proto_library` case, too, seems definitely wrong.

**Comment by achew22** [comment]
Out of curiosity, what happens if you put the proto in its own directory `path/to/dir/dogproto` where it has a `BUILD` file:

```
go_proto_library(
    name = "go_default_library",
    srcs = ["dog.proto"],
)
```

Does gazelle work for you then?

**Comment by jmhodges** [comment]
Yep, that was my workaround.

**Comment by achew22** [comment]
I haven't read through the gazelle code in a while, but if my memory is correct it doesn't do any parsing of BUILD files when trying to figure out what to depend on. I think it just shoots from the hip with what it thinks the name should be. It sees you `import "your.tld/path/to/dir/dogproto"` then it will put a dep on `//path/to/dir/dogproto:go_default_library` (assuming a go_prefix of "your.tld"). Based on a small amount of playing with golang's build system isn't this also what you would have to do to get it in a package of `your.tld/path/to/dir/dogproto` in it?

**Comment by jmhodges** [comment]
Yep!

Is that gazelle's design goal? If so, it'd be nice if gazelle made clear that it was expecting a more rigid bazel layout.

**Comment by achew22** [comment]
Unfortunately I'm an enthusiast not an author of the tool. I would be happy to contribute a patch or help you with the process. I don't have the commit bit though.

From [the docs](https://github.com/bazelbuild/rules_go/tree/fb1500ed88cc8d49f77b47dd0351275ebd4f4003/go/tools/gazelle#known-shortcomings) it has a note in the known shortcomings section that says:

 * bazel-style auto generating BUILD (where the library name is other than go_default_library)

If you were writing that document to a new user, how would you convey the message beyond that? I agree that it is a little spartan. Maybe we can wordsmith it together.

**Comment by jmhodges** [comment]
I think I'd like to hear from the maintainers before putting that work in. Thanks for the offer!

**Comment by jayconrod** [comment]
I'll volunteer to improve the Gazelle documentation. Agreed that it can be more descriptive.

Gazelle would need to read other BUILD files and Go sources in order to translate import paths to dependencies accurately. I'm not sure this would scale well.

**Comment by jmhodges** [comment]
That's cool! I guess I should open a ticket to make gazelle error in this case, then?

**Comment by jayconrod** [comment]
Reporting an error in this case would require gazelle to do more work than it currently does. Currently, it just reads the import paths from the .go source files then translates those into Bazel labels in a fairly deterministic fashion by assuming every library is named ":go_default_library". To report an error, it would need to read other BUILD files and confirm whether the target exists. Might need to check whether the other BUILD file was scheduled to be processed first and detect cyclic dependencies.

It would be nice to have this, but I'm not sure how soon we can get to it. Other features like cross-compilation and build tags need to be done first.

**Comment by jayconrod** [comment]
(Closing old Gazelle issues. If Gazelle isn't doing something you want, please file a new issue at https://github.com/bazelbuild/bazel-gazelle).

This should be working now that we have index-based dependency resolution. Gazelle will build an index of the libraries in a repository and their import paths, then use that to resolve deps. It should work with protos, too.
